sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -63,16 +74,345 @@ abstract class GpuShuffleBlockResolverBase(
   override def stop(): Unit = wrapped.stop()
 }
 
+/**
+ * The `ShuffleWriteMetricsReporter` is based on accumulators, which are not thread safe.
+ * This class is a thin wrapper that adds synchronization, since these metrics will be written
+ * by multiple threads.
+ * @param wrapped
+ */
+class ThreadSafeShuffleWriteMetricsReporter(wrapped: ShuffleWriteMetricsReporter)
+  extends ShuffleWriteMetrics {
+  override private[spark] def incBytesWritten(v: Long): Unit = synchronized {
+    wrapped.incBytesWritten(v)
+  }
+  override private[spark] def incRecordsWritten(v: Long): Unit = synchronized {
+    wrapped.incRecordsWritten(v)
+  }
+  override private[spark] def incWriteTime(v: Long): Unit = synchronized {
+    wrapped.incWriteTime(v)
+  }
+  override private[spark] def decBytesWritten(v: Long): Unit = synchronized {
+    wrapped.decBytesWritten(v)
+  }
+  override private[spark] def decRecordsWritten(v: Long): Unit = synchronized {
+    wrapped.decRecordsWritten(v)
+  }
+}
 
 object RapidsShuffleInternalManagerBase extends Logging {
   def unwrapHandle(handle: ShuffleHandle): ShuffleHandle = handle match {
     case gh: GpuShuffleHandle[_, _] => gh.wrapped
     case other => other
   }
+
+  /**
+   * A single-thread + queue used to compute a series of tasks serially.
+   * Used to compose a map of slots that can run independent of each other.
+   */
+  case class Task(body: () => Unit, handleException: Throwable => Unit)
+  private class Slot(i: Int) {
+    val tasks = new LinkedBlockingQueue[Task]()
+    val p = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+      .setNameFormat(s"rapids-shuffle-writer-$i")
+      .setDaemon(true)
+      .build)
+
+    def offer(task: Task): Unit = {
+      tasks.offer(task)
+    }
+    p.submit[Unit](() => {
+      while (true) {
+        var t: Task = null
+        try {
+          t = tasks.take()
+          t.body()
+        } catch {
+          case e: Throwable =>
+            t.handleException(e)
+        }
+      }
+    })
+
+    def shutdownNow(): Unit = p.shutdownNow()
+  }
+
+  // "slots" are a thread + queue thin wrapper that is used
```

Comment:
This comment seems like it should be a comment on the class itself rather than here.